# Chapitre 4 - Les fonctions


Une fonction est un bloc de code avec un but spécifique auquel vous pouvez donner un nom. Quand vous appelez cette fonction, vous exécutez le code qu’elle contient. Les fonctions vous laissent saisir des paramètres pour exécuter le même code sur différentes valeurs.  

Il y a différents types de fonctions dans Python :

Les fonctions intégrées fournies avec Python.
Les fonctions définies par l’utilisateur que les développeurs (vous !) créent. 
Vous avez déjà utilisé une fonction intégrée ! range() est une fonction dans laquelle vous pouvez saisir un nombre et elle renverra ou répondra avec une séquence allant de 0 à -1.

Pensez à une fonction comme à un moyen de réutiliser un ensemble d’instructions répétables. Vous la définissez avec le mot-clé def, le nom de la fonction, des parenthèses et deux-points:. Si la fonction a besoin d’un paramètre ou plus, il faut les saisir à l’intérieur des parenthèses, séparés par des virgules. Admettons que vous voulez ajouter deux nombres ensemble. L’extrait de code ci-dessous est une méthode ``add()`` qui prend deux nombres comme paramètres et renvoie la somme. 

```python
def add(x, y):
    return x + y
```
Vous pouvez appeler la fonction en utilisant son nom et en lui passant les valeurs que vous voulez additionner. Par exemple, si vous voulez additionner 2 et 3, vous pouvez faire ceci : 

```python
result = add(2, 3)
print(result)
```
Cela affichera 5. Vous pouvez également appeler la fonction avec des variables, comme ceci : 

```python
a = 2
b = 3
result = add(a, b)
print(result)
```

### Quand utiliser les fonctions et pourquoi ?

Quand on écrit beaucoup de code, on s’y perd et on fait des erreurs facilement entre les différentes fonctionnalités en cours. Les fonctions vous aident à séparer le code en sections plus petites. Comme ça vous gardez le fil sur ce que chaque partie est censée faire.

Vous en sortirez avec un code mieux écrit, mieux structuré et plus lisible.

Primordial de bien structurer son code, pour soi et aussi pour les personnes qui devront le reprendre après vous.

Plusieurs principes de programmation existent pour vous aider à écrire du code propre et lisible:

- Respectez les standards du code: Python a des conventions de nommage et de style (PEP 8) que vous devriez suivre pour rendre votre code cohérent et lisible.
    - Identation: 4 espaces par niveau d'indentation
    - Espaces autour des opérateurs: 1 espace avant et après les opérateurs binaires
    - Longueur des lignes: 79 caractères maximum
    - Nom de variables: Utilisez des noms significatifs et en minuscules, avec des underscores: snake_case pour séparer les mots (ex: `ma_variable`).

- Le principe du **"DRY" (Don't Repeat Yourself)** : Évitez de répéter le même code plusieurs fois. Utilisez des fonctions pour encapsuler le code réutilisable.

- Le principe de **responsabilité unique** : Chaque fonction doit avoir une seule responsabilité ou tâche. Cela rend le code plus facile à comprendre et à maintenir.

- **La documentation: commenter le code** : Utilisez des commentaires pour expliquer le but de chaque fonction et son fonctionnement. Cela aide les autres (et vous-même) à comprendre le code plus tard.
   - Maximum 72 caractères pour les commentaires & docstrings
   - Plusieurs conventions existent: https://stackoverflow.com/questions/3898572/what-are-the-most-common-python-docstring-formats
   - On verra dans le cours la convention ``reStructuredText (reST)``

- Annotations de type (Typage des variables): Depuis Python 3.5, vous pouvez annoter vos fonctions avec des types pour indiquer quel type de données elles attendent et renvoient. Cela peut aider à la lisibilité et à la détection des erreurs.

### La gestion des erreurs

Il est primordial de gérer les erreurs dans votre code. Vous pouvez utiliser des blocs ``try/except`` pour capturer les exceptions et éviter que le programme ne plante!

### Les tests unitaires
Les tests unitaires sont des morceaux de code qui vérifient si une fonction fonctionne comme prévu. Ils sont essentiels pour s'assurer que votre code est fiable et qu'il ne contient pas d'erreurs. Vous pouvez utiliser des bibliothèques comme `unittest` ou `pytest` pour écrire et exécuter des tests unitaires.

## Créer une fonction

Pour créer une fonction:
* On utilise le mot clef **def** qui signifie *define*
* Suivi du nom de la fonction
* De parenthèses pour les éventuels paramètres 
* Puis de la marque de début de bloc `:`  
* Tout ce qui est indenté après cette ligne de définition fait partie du code de la fonction
* La fonction peut renvoyer une valeur avec le mot clé `return` suivi de la valeur à renvoyer
* Si la fonction ne renvoie rien, elle renvoie `None` par défaut

Pour appeler la fonction on écrit son nom suivi de parenthèses `()`

In [29]:
def add_integers(a, b):
    # Identation: 4 espaces par niveau d'indentation
    # Longueur des lignes: 79 caractères maximum
    result = (
        a + b
    )  # Espaces autour des opérateurs: 1 espace avant et après les opérateurs binaire
    # Nom de variables: Utilisez des noms significatifs et en minuscules
    print("Addition de", a, "et", b, "est", result)
    return result


# Exemple d'utilisation de la fonction
add_integers(3, 5)
add_integers(10, 20)
add_integers(0, 0)
add_integers(-5, 5)

# On respecte ici les standards du code: Python a des conventions de nommage et de
# style (PEP 8):
# - Identation: 4 espaces par niveau d'indentation
# - Espaces autour des opérateurs: 1 espace avant et après les opérateurs binaires:
# - Longueur des lignes: 79 caractères maximum
# - Nom de variables: Utilisez des noms significatifs et en minuscules, avec des underscores: snake_case pour séparer les mots (ex: `ma_variable`).

Addition de 3 et 5 est 8
Addition de 10 et 20 est 30
Addition de 0 et 0 est 0
Addition de -5 et 5 est 0


0

### Le principe du DRY

Imaginons que vous devez calculer la moyenne de plusieurs listes de nombres. Au lieu de répéter le code de calcul de la moyenne pour chaque liste, vous pouvez créer une fonction réutilisable

In [30]:
# Sans DRY
liste1 = [1, 2, 3, 4, 5]
liste2 = [10, 20, 30, 40, 50]

# Calcul de la moyenne pour liste1
somme1 = sum(liste1)
moyenne1 = somme1 / len(liste1)

# Calcul de la moyenne pour liste2
somme2 = sum(liste2)
moyenne2 = somme2 / len(liste2)

print("Moyenne de liste1 :", moyenne1)
print("Moyenne de liste2 :", moyenne2)

Moyenne de liste1 : 3.0
Moyenne de liste2 : 30.0


In [31]:
# Avec DRY
def calculer_moyenne(liste):
    return sum(liste) / len(liste)


liste1 = [1, 2, 3, 4, 5]
liste2 = [10, 20, 30, 40, 50]

moyenne1 = calculer_moyenne(liste1)
moyenne2 = calculer_moyenne(liste2)

print("Moyenne de liste1 :", moyenne1)
print("Moyenne de liste2 :", moyenne2)

Moyenne de liste1 : 3.0
Moyenne de liste2 : 30.0


### La responsabilité unique

Supposons que vous avez une fonction qui lit un fichier, extrait des données, et calcule des statistiques. Pour respecter le principe de responsabilité unique, vous pouvez diviser cette fonction en plusieurs fonctions, chacune ayant une seule responsabilité.

In [36]:
# Ecrivons un fichier contenant 10 lignes
with open("fichier.txt", "w") as f:
    for i in range(10):
        f.write(f"{i + 1}\n")

In [38]:
# Sans principe de responsabilité unique
def traiter_fichier(fichier):
    with open(fichier, "r") as f:
        lignes = f.readlines()

    nombres = [int(ligne.strip()) for ligne in lignes]
    print("Liste des nombres :", nombres)
    somme = sum(nombres)
    moyenne = somme / len(nombres)

    print("Somme :", somme)
    print("Moyenne :", moyenne)


traiter_fichier("fichier.txt")

Liste des nombres : [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Somme : 55
Moyenne : 5.5


In [41]:
# Avec principe de responsabilité unique
def lire_fichier(fichier):
    with open(fichier, "r") as f:
        lignes = f.readlines()
    return [int(ligne.strip()) for ligne in lignes]


def calculer_somme(liste):
    return sum(liste)


def calculer_moyenne(liste):
    return sum(liste) / len(liste)


def afficher_resultats(liste):
    print("Liste des nombres :", liste)
    print("Somme :", calculer_somme(liste))
    print("Moyenne :", calculer_moyenne(liste))


def traiter_fichier(fichier):
    liste = lire_fichier(fichier)
    afficher_resultats(liste)


traiter_fichier("fichier.txt")

Liste des nombres : [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Somme : 55
Moyenne : 5.5


### Typer les arguments

Une fonction peut prendre n'importe quel type d'argument, c'est pourquoi
il est très important de "typer" les arguments.

In [28]:
def operation(a: int, b: int, op: str) -> int:
    print(f"a={a}, b={b}, op={op}")
    match op:
        case "+":
            return a + b
        case "-":
            return a - b
        case "*":
            return a * b
        case "/":
            return a / b
        case _:
            print("Opération non valide")


# Exemple d'utilisation de la fonction
print("Operation valide:", operation(10, 5, "+"))

# Exemple d'utilisation de la fonction avec une opération non valide
print(operation(10, 5, "%"))

# Exemple d'utilisation de la fonction avec des types incorrects
print("Operation avec types incorrects:", operation("10", "5", "/"))

# Python n'est pas typé statiquement, donc il n'y a pas de vérification de type
# au moment de la compilation.

a=10, b=5, op=+
Operation valide: 15
a=10, b=5, op=%
Opération non valide
None
a=10, b=5, op=/


TypeError: unsupported operand type(s) for /: 'str' and 'str'

### Paramètres positionnels ou nommés

Il est possible de passer les paramètres d'une fonction sans respecter l'ordre dans lequel ils ont été définis.

In [16]:
# Positionnels
operation(10, 5, "*")  # a = 10, b = 5, op = '*'

a=10, b=5, op=*


50

In [17]:
# Nommés
operation(a=10, b=5, op="-")  # a = 10, b = 5, op = '-'

# On peut, en nommant les arguments, changer l'ordre
operation(op="/", b=5, a=10)  # a = 10, b = 5, op = '/'

a=10, b=5, op=-
a=10, b=5, op=/


2.0

Ok, mais que se passe-t-il dans ce cas ?

In [ ]:
operation(6, a=5, op='*')

TypeError: operation() got multiple values for argument 'a'

Et dans celui-ci ?

In [ ]:
operation(a=5, 6, "*")

SyntaxError: positional argument follows keyword argument (2886018050.py, line 1)

**Explications**

Dans le cas `operation(6, a=5, op='*')` l'erreur se comprend assez bien: la valeur 6, paramètre positionnel, est affectée à `a` mais l'on précise aussi que `a` reçoit 5. Aussi *Python* est perdu et refuse de continuer.

Dans le second cas, l'erreur peut sembler moins évidente, mais elle repose sur une règle fondamentale en Python :

**Les paramètres positionnels doivent toujours précéder les paramètres nommés.**

### Pourquoi cette règle ?
Lorsque vous mélangez des paramètres positionnels et nommés, Python doit être capable d'associer chaque valeur à son paramètre correspondant. Si un paramètre nommé apparaît avant un paramètre positionnel, l'ordre devient ambigu et Python ne peut plus déterminer correctement à quel paramètre chaque valeur correspond.

## Documentation de la fonction

On peut aussi ajouter une **docstring** (documentation de la fonction) entre triple guillemets pour expliquer ce que fait la fonction. C'est une bonne pratique pour rendre le code plus lisible et compréhensible.

In [ ]:
def is_number_prime(n: int) -> bool:
    """Vérifie si un nombre est premier.
    Un nombre est premier s'il est supérieur à 1 et n'a pas de diviseur
    :param n: Le nombre à vérifier
    :return: True si le nombre est premier, False sinon
    """
    if n <= 1:
        return False
    for i in range(2, n):
        if n % i == 0:
            return False
    return True

On remarquera ici plusieurs choses:
* La docstring: documente la fonction
* Le typage des arguments: `n: int` signifie que `n` doit être un entier
* Le typage du retour: `-> int` signifie que la fonction renvoie un entier

Mais surtout:
* Les multiples return !

### Comment ca fonctionne ?

La fonction `is_number_prime` vérifie si un nombre est premier avec l'approche conditionnelle vu au chapitre précédent.
Si une condition est vérifiée, le bloc de code comprenant le `return` est exécuté et la fonction renvoie la valeur associée sans exécuter le reste du code!

Donc les multiples `return` permettent de sortir de la fonction à différents moments selon les conditions rencontrées.

In [23]:
print("Est-ce que 5 est un nombre premier?", is_number_prime(5))
print("Est-ce que 4 est un nombre premier?", is_number_prime(4))
print("Est-ce que 169 est un nombre premier?", is_number_prime(169))

Est-ce que 5 est un nombre premier? True
Est-ce que 4 est un nombre premier? False
Est-ce que 169 est un nombre premier? False


Python stocke cette chaîne dans une variable spéciale: l'attribut `__doc__` de la fonction.

Elle est utilisée pour générer l'aide en ligne

In [26]:
help(is_number_prime)

Help on function is_number_prime in module __main__:

is_number_prime(n: int) -> bool
    Vérifie si un nombre est premier.
    Un nombre est premier s'il est supérieur à 1 et n'a pas de diviseur
    :param n: Le nombre à vérifier
    :return: True si le nombre est premier, False sinon



In [27]:
print(is_number_prime.__doc__)

Vérifie si un nombre est premier.
    Un nombre est premier s'il est supérieur à 1 et n'a pas de diviseur
    :param n: Le nombre à vérifier
    :return: True si le nombre est premier, False sinon
    


## Exercice - Les suites mathématiques

L'objectif de cet exercice est de vous familiariser avec la création de fonctions en Python. Vous allez créer des fonctions pour calculer les termes de deux suites mathématiques célèbres : la suite de Fibonacci et la suite de Syracuse.

## Objectifs:
- Comprendre comment créer une fonction en Python
- Apprendre à utiliser des boucles et des conditions dans une fonction
- Apprendre à renvoyer des valeurs depuis une fonction
- Apprendre à documenter une fonction avec une docstring
- Apprendre à typer les arguments et le retour de la fonction
- Garder en tête le DRY & principe de responsabilité unique

### 1. La suite de Fibonacci
La suite de Fibonacci est une suite de nombres dans laquelle chaque nombre est la somme des deux précédents. La suite commence généralement par 0 et 1. Voici les premiers termes de la suite :
0, 1, 1, 2, 3, 5, 8, 13, 21, 34, ...

Elle s'écrit comme suit: u(n+2) = u(n+1) + u(n)

1. Ecrivez une fonction `fibonacci(n: int) -> int` qui prend un entier `n` en argument et renvoie le n-ième terme de la suite de Fibonacci. La fonction doit etre documentée avec une docstring et doit utiliser des annotations de type.
2. Ajouter une fonction `ecrire_resultat(n: int) -> None` qui prend un entier `n` en argument et qui appelle la fonction `fibonacci(n)` pour écrire le résultat dans un fichier texte `resultat.txt`: `Le résultat de fibonacci(n) est : <resultat>`.
3. Ajouter une fonction `lire_resultat() -> None` qui lit le fichier `resultat.txt` et affiche le contenu à l'écran.
4. Ajouter une fonction `main()` qui appelle les fonctions `ecrire_resultat(n)` et `lire_resultat()`. La fonction `main()` doit être appelée si le script est exécuté directement.

### Si vous avez déjà fini:

### 2. La suite de Syracuse

La suite de Syracuse est une suite de nombres définie de la manière suivante :
- Si n est pair, le terme suivant est n/2
- Si n est impair, le terme suivant est 3n + 1
- La suite se termine lorsque le terme atteint 1.

1. Ecrivez une fonction `syracuse_sequence(n: int) -> List[int]` qui prend un entier `n` en argument et renvoie la suite de Syracuse sous forme de liste. La fonction doit être documentée avec une docstring et doit utiliser des annotations de type.
2. Ecrivez une fonction `syracuse_convergence_index(n: int) -> int` qui prend un entier `n` en argument et renvoie l'index du premier terme de la suite de Syracuse qui atteint 1. La fonction doit être documentée avec une docstring et doit utiliser des annotations de type.

In [ ]:
# A vous de jouer!

## La gestion des erreurs

Il est primordial de gérer les erreurs dans votre code. Vous pouvez utiliser des blocs `try/except` pour capturer les exceptions et éviter que le programme ne plante!

Il est possible également de lever des exceptions personnalisées avec le mot clé `raise`. Cela peut être utile pour signaler des erreurs spécifiques dans votre code. Par exemple, si vous avez une fonction qui attend un entier positif, vous pouvez lever une exception si l'utilisateur entre un nombre négatif.

In [42]:
def divide(a: float, b: float) -> float:
    """Divise a par b.
    :param a: Le numérateur
    :param b: Le dénominateur
    :return: Le résultat de la division
    :raises ZeroDivisionError: Si b est égal à 0
    """
    if b == 0:
        raise ZeroDivisionError("Attention: le dénominateur ne peut pas être zéro.")
    return a / b

In [45]:
divide(10, 2)  # 5.0

5.0

In [44]:
divide(10, 0)  # Lève une exception ZeroDivisionError

ZeroDivisionError: Attention: le dénominateur ne peut pas être zéro.

In [46]:
# Ou encore avec try/except
def divide(a: float, b: float) -> float:
    """Divise a par b.
    :param a: Le numérateur
    :param b: Le dénominateur
    :return: Le résultat de la division
    :raises ZeroDivisionError: Si b est égal à 0
    """
    try:
        return a / b
    except ZeroDivisionError:
        print("Erreur: le dénominateur ne peut pas être zéro.")
        return None

In [47]:
divide(10, 0)  # 5.0

Erreur: le dénominateur ne peut pas être zéro.


In [ ]:
# Il est possible de capturer une exception "Générale" avec Exception, mais ce n'est pas une bonne pratique
# car cela peut masquer d'autres erreurs.
# Il est préférable de capturer des exceptions spécifiques.

In [ ]:
def divide(a: int, b: int) -> float:
    """Divise a par b.
    :param a: Le numérateur
    :param b: Le dénominateur
    :return: Le résultat de la division
    """
    try:
        return a / b
    except Exception:
        print("Erreur: Il n'est pas possible de diviser par 0.")
        return None

In [52]:
divide(10, 0)

Erreur: Il n'est pas possible de diviser par 0.


In [53]:
divide("10", 0)

Erreur: Il n'est pas possible de diviser par 0.


In [59]:
# Il est possible de lever plusieurs exceptions & d'utiliser la syntaxe "as" pour capturer l'exception et l'afficher
def divide(a: int, b: int) -> float:
    """Divise a par b.
    :param a: Le numérateur
    :param b: Le dénominateur
    :return: Le résultat de la division
    """
    try:
        return a / b
    except ZeroDivisionError as e:
        print(f"Erreur: {e}, l'opération a échoué.")
        return None
    except TypeError as e:
        print(f"Erreur: {e}, l'opération a échoué.")
        return None

In [60]:
divide(10, 0)

Erreur: division by zero, l'opération a échoué.


In [61]:
divide("10", 0)

Erreur: unsupported operand type(s) for /: 'str' and 'int', l'opération a échoué.


## Exercice 🐍 - Gestion des Exceptions dans les Fonctions
L'objectif de cet exercice est de vous familiariser avec la gestion des exceptions dans les fonctions Python. Vous allez créer des fonctions qui traite un email reçu et générer un dictionnaire à partir de celui-ci. Vous allez également apprendre à gérer les exceptions potentielles qui peuvent survenir lors de l'exécution de ces fonctions.

## Objectifs:
- Apprendre à gérer les exceptions dans une fonction
- Apprendre à utiliser des blocs try, except, et finally

### Manipulation de Chaînes de Caractères avec Gestion des Exceptions
Vous allez créer une fonction qui traite un email reçu et gère les exceptions potentielles, comme des formats incorrects ou des valeurs manquantes.

1. Écrivez une fonction traiter_email(email: str) -> dict qui prend une chaîne de caractères représentant un email en argument et renvoie un dictionnaire contenant les informations extraites.

Le format d'un email est le suivant :
```
"Sujet": <sujet>\n
"De": <expéditeur>\n
"Contenu": <contenu>
```
Tips:
* Pensez à split('\n') pour créer une liste de lignes !
* Pensez à split(':') pour séparer les clés et les valeurs !

La fonction doit gérer les exceptions suivantes :

* ValueError : Si l'email ne contient pas les informations attendues.
* TypeError : Si l'argument n'est pas une chaîne de caractères.
* Exception : Pour toute autre exception inattendue.

La fonction doit être documentée avec une docstring et utiliser des annotations de type.

2. Ajoutez une fonction valider_dictionnaire(donnees: dict) -> bool qui prend un dictionnaire en argument et vérifie s'il contient certaines clés obligatoires (par exemple, "sujet", "expediteur", et "contenu"). La fonction doit gérer les exceptions suivantes :

* KeyError : Si une clé obligatoire est manquante.
* TypeError : Si l'argument n'est pas un dictionnaire.
* Exception : Pour toute autre exception inattendue.

3. Ajoutez une fonction main() qui appelle les fonctions traiter_email et valider_dictionnaire avec des exemples de données. La fonction main() doit être appelée si le script est exécuté directement.

In [ ]:
email = """
Sujet: Réunion de projet MIMO
De: louis@giron-dom.eu
Contenu: Bonjour à tous, la réunion de projet MIMO est prévue pour demain à 10h!*
"""

# A vous de jouer !

## Tests unitaires

In [26]:
# Ecrire un test unitaire est primordial pour s'assurer que le code fonctionne comme prévu.
# Développement TDD: Test driven development


def suite_de_conway(n: int) -> list:
    """La suite de Conway est une suite d'entiers étudiée en 1986 par le mathématicien britannique John Horton Conway,
    initialement sous le nom de « suite audioactive ». Elle est également connue sous le nom anglais de Look and Say
    sequence (« suite regarde et dis »). Elle est répertoriée comme suite A005150 de l'OEIS.
    Dans cette suite autoréférente, chaque terme se détermine en annonçant les chiffres formant le terme précédent.

    :param n: Le nombre de termes à générer
    :return: La suite de Conway sous forme de liste

    Exemple:
    X_0 = 1
    X_1 = 11 (X_0 est égal à 1 "1")
    X_2 = 21 (X_1 est égal à 2 "1")
    X_3 = 1211 (X_2 est égal à 1 "2" et 1 "1")
    X_4 = 111221 (X_3 est égal à 1 "1", 1 "2" et 2 "1")
    ...
    """
    x_0 = 1
    seq = [x_0]

    for i in range(1, n + 1):
        # On convertit le nombre en chaîne de caractères pour le traiter
        previous_term = str(seq[i - 1])
        next_term = ""

        # On compte les occurrences de chaque chiffre consécutif: initialisation à 1
        count = 1
        for j in range(1, len(previous_term)):
            # On compare le chiffre actuel avec le précédent
            # Si ils sont égaux, on incrémente le compteur
            if previous_term[j] == previous_term[j - 1]:
                count += 1
            else:
                # Sinon, on ajoute le compteur et le chiffre précédent à la chaîne
                next_term += f"{count}{previous_term[j - 1]}"
                count = 1

        # On ajoute le dernier groupe de chiffres (e.g. "11" ou "22")
        next_term += f"{count}{previous_term[-1]}"
        seq.append(int(next_term))
    return seq

In [27]:
suite_de_conway(5)

[1, 11, 21, 1211, 111221, 312211]

L'implémentation de la suite de Conway n'est pas aisé ! Nous ne sommes pas certains de l'avoir correctement implémenté !

Pour se faire, il est possible d'utiliser un test unitaire. Un test unitaire est un morceau de code qui teste une fonction ou une méthode pour s'assurer qu'elle fonctionne comme prévu. Python a une bibliothèque intégrée appelée unittest qui vous permet d'écrire et d'exécuter des tests unitaires.

In [28]:
# Ici le mot clef est "assert"
def test_suite_de_conway():
    """Tests pour la fonction suite_de_conway
    Valeurs attendues: https://fr.wikipedia.org/wiki/Suite_de_Conway
    """
    assert suite_de_conway(0) == [1]
    assert suite_de_conway(1) == [1, 11]
    assert suite_de_conway(2) == [1, 11, 21]
    assert suite_de_conway(3) == [1, 11, 21, 1211]
    assert suite_de_conway(4) == [1, 11, 21, 1211, 111221]
    assert suite_de_conway(5) == [1, 11, 21, 1211, 111221, 312211]
    assert suite_de_conway(6) == [1, 11, 21, 1211, 111221, 312211, 13112221]

In [ ]:
test_suite_de_conway()
# Le test passe !

In [31]:
def wrong_addition(a: int, b: int) -> int:
    """wrong implemntation of an addition"""
    return a - b


def test_wrong_addition():
    """Tests pour la fonction wrong_addition"""
    actual_1 = wrong_addition(2, 3)
    expected_1 = 5

    assert actual_1 == expected_1, (
        f"Nous devrions obtenir 2 + 3 = {expected_1}, nous avons obtenu {actual_1}"
    )


test_wrong_addition()

AssertionError: Nous devrions obtenir 2 + 3 = 5, nous avons obtenu -1

## Fonctions récursives

Une fonction récursive est une fonction qui s'appelle elle-même pour résoudre un problème. Cela peut être utile pour des problèmes qui peuvent être divisés en sous-problèmes similaires.

Il y a un point important dans l'écriture de ces fonctions, il faut une condition pour qu'elles cessent de s'appeler à un moment donné, sinon elles boucleront à l'infini

Exemples:
- Compte à rebours
- Fibonacci ...


In [36]:
import time

def compte_a_rebours(n: int) -> None:
    """Compteur à rebours de n à 0
    :param n: Le nombre de départ
    :return: None
    """
    for i in range(n, 0, -1):
        print(i)
        time.sleep(0.5)  # Pause d'une seconde entre chaque nombre
    print("Partez!")

compte_a_rebours(5)

5
4
3
2
1
Partez!


Mais nous pourrions aussi jouer sur le fait qu'elle affiche un nombre puis le suivant.

Pourquoi ne pas lui demander de se rappeler elle-même avec le nombre suivant pour l'afficher à son tour ?

In [49]:
def compte_a_rebours(n: int) -> None:
    """Compteur à rebours de n à 0
    :param n: Le nombre de départ
    :return: None
    """
    print(n)
    #time.sleep(0.5)  # Pause d'une seconde entre chaque nombre
    compte_a_rebours(n -1)

compte_a_rebours(5)

5
4
3
2
1
0
-1
-2
-3
-4
-5
-6
-7
-8
-9
-10
-11
-12
-13
-14
-15
-16
-17
-18
-19
-20
-21
-22
-23
-24
-25
-26
-27
-28
-29
-30
-31
-32
-33
-34
-35
-36
-37
-38
-39
-40
-41
-42
-43
-44
-45
-46
-47
-48
-49
-50
-51
-52
-53
-54
-55
-56
-57
-58
-59
-60
-61
-62
-63
-64
-65
-66
-67
-68
-69
-70
-71
-72
-73
-74
-75
-76
-77
-78
-79
-80
-81
-82
-83
-84
-85
-86
-87
-88
-89
-90
-91
-92
-93
-94
-95
-96
-97
-98
-99
-100
-101
-102
-103
-104
-105
-106
-107
-108
-109
-110
-111
-112
-113
-114
-115
-116
-117
-118
-119
-120
-121
-122
-123
-124
-125
-126
-127
-128
-129
-130
-131
-132
-133
-134
-135
-136
-137
-138
-139
-140
-141
-142
-143
-144
-145
-146
-147
-148
-149
-150
-151
-152
-153
-154
-155
-156
-157
-158
-159
-160
-161
-162
-163
-164
-165
-166
-167
-168
-169
-170
-171
-172
-173
-174
-175
-176
-177
-178
-179
-180
-181
-182
-183
-184
-185
-186
-187
-188
-189
-190
-191
-192
-193
-194
-195
-196
-197
-198
-199
-200
-201
-202
-203
-204
-205
-206
-207
-208
-209
-210
-211
-212
-213
-214
-215
-216
-217
-218
-219
-

RecursionError: maximum recursion depth exceeded

Lorsque vous exécutez le code ci-dessus, nous obtenons une erreur **Recursion error**

Le problème est que la fonction se rappelle à l'infini.

Mais la mémoire de l'ordinateur est limitée en taille et il ne peut pas gérer une infinité d'appels.

Il convient de lui préciser quand s'arrêter...

In [54]:
def compte_a_rebours_recursive(n: int) -> None:
    """Compteur à rebours de n à 0
    :param n: Le nombre de départ
    :return: None
    """
    if n > 0:
        print(n)
        time.sleep(0.5)  # Pause d'une seconde entre chaque nombre
        compte_a_rebours_recursive(n -1)
    else:
        print("Partez!") # Condition d'arrêt de la récursion (n == 0)

compte_a_rebours_recursive(5)

5
4
3
2
1
Partez!


In [59]:
# Autre exemple, la suite de Fibonacci
# On a vu dans l'exercice précédent que la suite de Fibonacci est définie par
# la relation de récurrence suivante:
# u(n) = u(n-1) + u(n-2)
# avec u(0) = 0 et u(1) = 1

def fibonacci_recursive(n: int) -> int:
    """Calcul la somme de la suite de Fibonacci de manière récursive.
    :param n: Le rang de la suite de Fibonacci
    :return: La valeur de la suite de Fibonacci à ce rang
    """
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fibonacci_recursive(n - 1) + fibonacci_recursive(n - 2)
    
# Exemple d'utilisation de la fonction
for i in range(10):
    print(f"Fibonacci({i}) = {fibonacci_recursive(i)}")

Fibonacci(0) = 0
Fibonacci(1) = 1
Fibonacci(2) = 1
Fibonacci(3) = 2
Fibonacci(4) = 3
Fibonacci(5) = 5
Fibonacci(6) = 8
Fibonacci(7) = 13
Fibonacci(8) = 21
Fibonacci(9) = 34


## Exercice 🐍 - Calcul de la factorielle d'un nombre récursivement

L'objectif de cet exercice est de vous familiariser avec la création de fonctions récursives en Python. Vous allez créer une fonction qui calcule la factorielle d'un nombre entier positif et la tester !

## Objectifs:
- Comprendre comment créer une fonction récursive en Python
- Apprendre à utiliser des conditions dans une fonction récursive
- Apprendre à renvoyer des valeurs depuis une fonction récursive
- Apprendre à tester une fonction

### La factorielle d'un nombre

La factorielle d'un nombre entier positif n est le produit de tous les entiers positifs inférieurs ou égaux à n. Elle est notée n! et se définit comme suit :
- n! = n × (n - 1)! pour n > 0
- 0! = 1

donc 5! = 1 * 2 * 3 * 4 * 5

1. Écrivez une fonction `factorielle(n: int) -> int` qui prend un entier positif `n` en argument et renvoie la factorielle de `n`. La fonction doit être documentée avec une docstring et utiliser des annotations de type.
2. Écrivez une fonction `factorielle_recursive(n: int) -> int` qui utilise la récursivité pour calculer la factorielle de `n`. La fonction doit être documentée avec une docstring et utiliser des annotations de type.
3. Écrivez un test unitaire qui vérifie que la fonction `factorielle_recursive(n)` renvoie le bon résultat pour plusieurs valeurs de `n`. Vous testerez pour cela:
* Factoriel de 0 == 1
* Factoriel de 1 == 1
* Factoriel de 3 == 6
* Factoriel de 5 == 120

In [ ]:
# A vous de jouer